Docsp 9569 (#53)
mongodb/docs-java: source/fundamentals/aggregation.txt

===========
Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Aggregation operations process data and return computed results, modeled on the idea of :wikipedia:`unix pipelines </Pipeline_(Unix)?oldid=993313477>`.
The :manual:`aggregation pipeline </core/aggregation-pipeline/>` can be thought of as a factory assembly line. Along the
line are assembly stations, called :manual:`stages </meta/aggregation-quick-reference/#stages>`. Like an assembly station
configured with specific tooling to perform its task, each ``stage`` uses
:manual:`operator expressions </meta/aggregation-quick-reference/#operator-expressions>` to perform some operation or
combination of operations.

In a car factory, for example, one station might shape and bend the frame using a hydraulic press and a welder, another
might affix the doors with a drill, and so on.

Find and Aggregation Comparison
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Using ``find`` operations, you can:

- select **what** documents to return

> - select **what** fields to return

Author: I had gone back and forth on this since italics can be missed, but they are probably more appropriate here. I want to leave the emphasis on the current words as is as that's idea.

- sort the results

Using ``aggregation`` operations, you can:

- perform all ``find`` operations
- calculate fields
- summarize and group values
- reshape documents

Illustrated Example
~~~~~~~~~~~~~~~~~~~

In the animation below, every document in the collection enters the pipeline and is first processed by
the ``$match`` stage. In this example, the stage selectively matches documents that are either red or blue. The documents
continue flowing through the pipeline and enter the ``$project`` stage which transforms them from squares to circles.
Lastly, the documents flow into the ``$group`` stage, configured to produce the ratio of red to blue circles.

.. cssclass:: bordered-figure
.. figure:: /figures/agg_factory.gif
   :alt: Animation showing how the aggregation pipeline works, conceptually.

Using the :ref:`Builders <builders>` API, the same operation may look something like:

.. code-block:: java

   Arrays.asList(

>        Aggregates.match(Filters.in("color", "red", "blue")), // only allow red and blue shapes to pass

Author: It is 4-space indentation, and shows up as such for me. I'll switch to a literalinclude to see if that helps

.. code-block:: java

   import com.mongodb.client.MongoClient;
   import com.mongodb.client.MongoClients;
   import com.mongodb.client.MongoCollection;
   import com.mongodb.client.MongoDatabase;
   import com.mongodb.client.model.Accumulators;
   import com.mongodb.client.model.Aggregates;
   import com.mongodb.client.model.Filters;
   import org.bson.Document;

   import java.util.Arrays;

Connect to a MongoDB Deployment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Connect to a MongoDB deployment and declare and define ``MongoDatabase`` and ``MongoCollection`` instances.

For example, include the following code to connect to a standalone MongoDB deployment running on localhost on port
``27017`` and define ``database`` to refer to the ``test`` database and ``collection`` to refer to the ``restaurants``
collection.

.. code-block:: java

   public class AggTourOne {

     public static void main(String[] args) {
       MongoClient mongoClient = MongoClients.create();
       MongoDatabase database = mongoClient.getDatabase("test");
       MongoCollection<Document> collection = database.getCollection("restaurants");

> 

Reviewer: Just wanted to make sure this was visible since it looks like you gave it thumbs, but doesn't look like any changes were made.

- Uses  a :manual:`$group <reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``
  field, accumulating a count of documents for each distinct value of ``stars``. The example uses
  `Aggregates.group <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#group(org.bson.conversions.Bson)>`__
  to build the ``$group`` stage and `Accumulators.sum <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators#sum(java.lang.String,TExpression)>`__
  to build the accumulator expression. The :manual:`accumulator expressions <reference/operator/aggregation/#accumulators-group>`
  used within the ``$group`` stage can be built with the `Accumulators <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators.html>`__
  helper class.

.. code-block:: java

   collection.aggregate(
         Arrays.asList(
                 Aggregates.match(Filters.eq("categories", "Bakery")),
                 Aggregates.group("$stars", Accumulators.sum("count", 1))
         )

>    ).forEach(printBlock);

Reviewer: I expected to see output after all of the examples that included printBlock. You might consider showing sample output or removing the forEach call to make that apparent.

Use Aggregation Expressions
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Java driver provides builders for :java-sync-api:`accumulator expressions<com/mongodb/client/model/Accumulators.html>`

> for use with :java-sync-api:`$group <com/mongodb/client/model/Aggregates.html#group(TExpression,java.util.List)>`. All
> other expressions must be manually built.

Reviewer: Nitpick: I would prefer to be specific as to who performs what rather than mentioning the end-state. I think this helps avoid passive voice too.  In this sentence, we could reword "X must be [done]" as something like:

```suggestion
for use with :java-sync-api:`$group <com/mongodb/client/model/Aggregates.html#group(TExpression,java.util.List)>`. For all other expressions, you must manually declare the operators in JSON or compatible document format.
```

In the following example, the aggregation pipeline uses a
`$project <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#project(org.bson.conversions.Bson)>`__
stage and various
`Projections <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Projections.html>`__

> to return only the ``name`` field, and the calculated field ``firstCategory`` whose value is the first element in the

Reviewer: Would it be worthwhile linking to the "arrayElemAt" aggregation operator (in an API links section at the end)? Also, I wonder if it would be helpful to explain the "$" field path for categories.

===========
Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Aggregation operations process data in your MongoDB collections and return computed results.

> A powerful way to think of aggregation is like a car factory. Within the car factory is an assembly line, along which

Reviewer 2: Editorial nit: I would not claim powerful.  That's in the eye of the beholder, and may turn off some readers who don't find it powerful, or are still confused.  Perhaps just "One way to think of..."

Author: Great point, thanks!

   public class AggTour {

       public static void main(String[] args) {
           // Replace the uri string with your MongoDB deployment's connection string
           String uri = "mongodb+srv://<user>:<password>@<cluster-url>?retryWrites=true&w=majority";

           MongoClient mongoClient = MongoClients.create(uri);
           MongoDatabase database = mongoClient.getDatabase("aggregation");
           MongoCollection<Document> collection = database.getCollection("restaurants");

           // aggregation here
       }
   }

.. tip::

   For information on connecting to MongoDB, see the :ref:`Connection Guide <mongoclient>`

Insert the Data
+++++++++++++++

.. literalinclude:: /includes/fundamentals/code-snippets/AggTour.java
   :language: java
   :dedent:
   :start-after: begin insert
   :end-before: end insert

Basic Aggregation Example
~~~~~~~~~~~~~~~~~~~~~~~~~

To perform an aggregation, pass a list of aggregation stages to the `MongoCollection.aggregate() <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/MongoCollection.html#aggregate(com.mongodb.client.ClientSession,java.util.List,java.lang.Class)>`__
method.

> The Java driver provides the `Aggregates <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html>`__

Reviewer 2: I feel like we should make clear somewhere that it's not required to use the builders.  Not sure whether this is the place though.  What do you think?

Author: I was going to frame this in the builders page, and frame it as "here's this manually constructed aggregation, here's the builder version".

===========
Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Aggregation operations process data in your MongoDB collections and return computed results. The **aggregation pipeline**
is like a factory assembly line. Along the line are assembly stations, called **aggregation stages** . Like an assembly
station configured with specific tooling to perform its task, each ``stage`` uses **expressions** to perform some

> operation or combination of operations.

Reviewer: The sentences switch between extended metaphor and simile which makes it a little more difficult to read. I would stick to one of the two throughout the paragraph. E.g.

Extended metaphor:
The aggregation pipeline is a factory assembly line.  The sequentially-arranged assembly stations are aggregation stages. Each assembly station at which a specific task is performed is an aggregation stage with a specific operation or combination of operations. 

Simile (compare two things):
Aggregation pipelines are like a factory assembly line. Aggregation stages are like the assembly stations that sequentially put together the finished product. Similar to assembly stations, aggregation stages use specific tools called expressions to perform a specific operation or combination of operations.

> Aggregation Does More

Reviewer: Nitpick: This should probably specify what you're comparing Aggregation to here. E.g.
"Aggregation is More Flexible than the Find Operation"
Currently, it doesn't match the specificity and organization of information of the rest of the documentation. E.g. sections that describe what it is, show how you use it, or what you need to set up to use it

Author: This was meant to stand out.

~~~~~~~~~~~~~~~~~~~~~

Using ``find`` operations, you can:

- select *what* documents to return
- select *what* fields to return
- sort the results

Using ``aggregation`` operations, you can:

> - perform all ``find`` operations

Reviewer: I think we should take the approach of being objective and list the drawbacks as well such as memory constraints rather than advocating only for aggregation over find.

- rename fields
- calculate fields
- summarize data
- group values

Illustrated Example
~~~~~~~~~~~~~~~~~~~

In a car factory assembly line, one station might shape and bend the frame using a hydraulic press and a welder, another
might affix the doors with a drill, and so on.

In the following animation, every document in the collection enters the pipeline and is first filtered by
the ``$match`` stage. In this example, the stage selectively matches documents that are either "red" or "blue",
filtering "green" documents out. The documents continue flowing through the pipeline and enter the ``$project`` stage,
which transforms them from squares to circles. Lastly, the documents flow into the ``$group`` stage, configured to
produce the ratio of red to blue circles.

.. figure:: /figures/agg_factory.gif
   :border:
   :alt: Animation showing how the aggregation pipeline works, conceptually.

Using the **Aggregation builders** API, the same operation may look something like:

.. literalinclude:: /includes/fundamentals/code-snippets/agg-factory.java
   :language: java

.. code-block:: json

   {"_id": 0, "red": 3, "blue": 2}

> Terms

Reviewer: Would it be possible to add a sentence under this heading to explain that these are links at which the ? We don't have any similarly-named sections across docs-java so the purpose of this section might not be easily understood.

~~~~~

- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`
- :ref:`Aggregation Builders <builders>`

> Runnable Examples

Reviewer: Would you consider adding sub-headings under this for "Setup" and then "Aggregation Example"? When I read this heading, I expect to see the sample code rather than the initial setup so it could be helpful to identify what's happening in each paragraph.

-----------------

The following examples require a ``restaurants`` collection in the ``test`` database. To create and populate the collection,
follow the `directions <https://github.com/mongodb/docs-assets/tree/drivers>`_ to download and load the dataset. If you

> prefer to load the data directly and have both ``mongoimport`` and ``curl`` in your path:

Reviewer: ```suggestion
prefer to load the data directly and have both ``mongoimport`` and ``curl`` in your path, you can run the following instead:
```

.. tabs::

   .. tab::
      :tabid: Atlas

>       .. code-block:: sh

Author: thanks for the info

>          mongoimport --uri mongodb+srv://<your-atlas-uri>/test --collection restaurants --drop --file <(curl https://raw.githubusercontent.com/mongodb/docs-assets/drivers/restaurants.json)

Reviewer: Nitpick: what happens here when curl fails? Maybe we can instead split it into two commands so that the user would see the related error from curl before attempting mongoimport? Same comment for localhost tab.

Author: I'm not convinced the value separating the two would provide in the unlikely scenario the curl fails for a small 10 document demo collection is worth worrying over. Open to more discussion on this in private channels.

   import com.mongodb.client.MongoCollection;
   import com.mongodb.client.MongoDatabase;
   import com.mongodb.client.model.Accumulators;
   import com.mongodb.client.model.Aggregates;
   import com.mongodb.client.model.Filters;
   import org.bson.Document;

   import java.util.Arrays;

Connect to a MongoDB Deployment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Connect to a MongoDB deployment and declare and define ``MongoDatabase`` and ``MongoCollection`` instances.

For example, include the following code to connect to a standalone MongoDB deployment running on localhost on port
``27017`` and define ``database`` to refer to the ``test`` database and ``collection`` to refer to the ``restaurants``
collection.

.. code-block:: java

   public class AggTourOne {

     public static void main(String[] args) {
       MongoClient mongoClient = MongoClients.create();
       MongoDatabase database = mongoClient.getDatabase("test");
       MongoCollection<Document> collection = database.getCollection("restaurants");

       }
   }

.. tip::

   For information on connecting to MongoDB, see the :ref:`Connection Guide <mongoclient>`

Perform Aggregation
~~~~~~~~~~~~~~~~~~~

> To perform an aggregation, pass a list of :manual:`aggregation stages </meta/aggregation-quick-reference/#stages>` to

Reviewer: If this is under the Terms section, does it need to be linked again?

- Uses  a :manual:`$group <reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``
  field, accumulating a count of documents for each distinct value of ``stars``.

  The example uses `Aggregates.group <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#group(org.bson.conversions.Bson)>`__
  to build the ``$group`` stage and `Accumulators.sum <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators#sum(java.lang.String,TExpression)>`__
  to build the accumulator expression. The :manual:`accumulator expressions <reference/operator/aggregation/#accumulators-group>`
  used within the ``$group`` stage are built with the `Accumulators <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators.html>`__
  helper class.

.. code-block:: java

   collection.aggregate(
         Arrays.asList(
                 Aggregates.match(Filters.eq("categories", "Bakery")),
                 Aggregates.group("$stars", Accumulators.sum("count", 1))
         )
   ).forEach(printBlock);

> 

Reviewer: I like to separate the output from sample code by explaining that running the code should produce the following output. Random example (authored by Katie) https://github.com/mongodb/docs-java/blob/master/source/fundamentals/crud/read-operations/limit.txt#L67
Would you agree with this or otherwise, is it not a good idea?

Author: I felt it was self-evident, but perhaps that's an incorrect position

.. code-block:: json

> 

Reviewer: Just to clean up the UI:
```suggestion
    :copyable: false
```

   {"_id": 4, "count": 2}
   {"_id": 5, "count": 1}

Use Aggregation Expressions
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Java driver provides builders for :java-sync-api:`accumulator expressions<com/mongodb/client/model/Accumulators.html>`
for use with :java-sync-api:`$group <com/mongodb/client/model/Aggregates.html#group(TExpression,java.util.List)>`. You
must declare all other expressions in JSON format or compatible document format.

> .. example::

Reviewer: To me, the formatting is confusing when using this in the "Runnable Examples" section -- this example is within the "Example" box, but the other examples are not. I wonder if we should identify each example with a sub-heading that includes "Example" similar to Kafka docs here https://docs.mongodb.com/kafka-connector/current/kafka-sink-postprocessors.

   The following two lines of code produce the same expression, :manual:`$arrayElemAt </reference/operator/aggregation/arrayElemAt/>`.
   The ``$`` in front of **categories** tells MongoDB that this is a :manual:`field path </meta/aggregation-quick-reference/#expressions>`,
   using the **categories** field from the input document.

   .. code-block:: java

      new Document("$arrayElemAt", Arrays.asList("$categories", 0))

   .. code-block:: java

      Document.parse("{ $arrayElemAt: ['$categories', 0] }")

> In the following example, the aggregation pipeline uses a

Reviewer: Would it be helpful to explain all the Projections static functions called or provide links to the api docs for them?

Author: I don't think so in this case. I think it will we be more user friendly to ultimately link to the builders pages we create as part of the docs site.

Aggregation
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

Aggregation operations process data in your MongoDB collections and return computed results.

A powerful way to think of aggregation is like a car factory. Within the car factory is an assembly line, along which
are assembly stations with specialized tools to do a specific job, like drills and welders. Raw parts enter the factory,
which are then transformed and assembled into a finished product.

The **aggregation pipeline** is the assembly line, **aggregation stages** are the assembly stations, and
**operator expressions** are the specialized tools.

Aggregation and Find Operations Compared
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Using ``find`` operations, you can:

- select *what* documents to return
- select *what* fields to return
- sort the results

Using ``aggregation`` operations, you can:

- perform all ``find`` operations
- rename fields
- calculate fields
- summarize data
- group values

Aggregation operations have some :manual:`limitations <core/aggregation-pipeline-limits/>` you must keep in mind:

> - Returned documents must not violate the :manual:`BSON Document Size limit </reference/limits/#BSON-Document-Size>`,
>   currently 16 megabytes.

Reviewer: Thought it was strange to mix title case and normal case within the same term. Also thought "currently" signaled that it's a value that changes often.

```suggestion
- Returned documents must not violate the :manual:`BSON document size limit </reference/limits/#BSON-Document-Size>`
  of 16 megabytes.
```

filtering "green" documents out. The documents continue flowing through the pipeline and enter the ``$project`` stage,
which transforms them from squares to circles. Lastly, the documents flow into the ``$group`` stage, configured to
produce the ratio of red to blue circles.

.. figure:: /figures/agg_factory.gif
   :border:
   :alt: Animation showing how the aggregation pipeline works, conceptually.

Using the **Aggregation builders** API, the same operation may look something like:

.. literalinclude:: /includes/fundamentals/code-snippets/agg-factory.java
   :language: java
   :copyable: false

Which will produce output similar to the following:

.. code-block:: none
   :copyable: false

   {"_id": 0, "red": 3, "blue": 2}

Useful Links
~~~~~~~~~~~~

- :manual:`Aggregation pipeline </core/aggregation-pipeline/>`
- :manual:`Aggregation stages </meta/aggregation-quick-reference/#stages>`
- :manual:`Operator expressions </meta/aggregation-quick-reference/#operator-expressions>`
- :ref:`Aggregation Builders <builders>`

Runnable Examples
-----------------

The following examples require a ``restaurants`` collection in the ``test`` database. To create and populate the collection,

> follow the `directions <https://github.com/mongodb/docs-assets/tree/drivers>`_ to download and load the dataset. If you

Reviewer: ```suggestion
follow `these directions <https://github.com/mongodb/docs-assets/tree/drivers>`_ to download and load the dataset. If you
```

      :tabid: Atlas

      .. code-block:: console

         curl -O https://raw.githubusercontent.com/mongodb/docs-assets/drivers/restaurants.json
         mongoimport --uri mongodb+srv://<your-atlas-uri>/test --collection restaurants --drop --file restaurants.json

   .. tab::
      :tabid: localhost

      .. code-block:: console

         curl -O https://raw.githubusercontent.com/mongodb/docs-assets/drivers/restaurants.json
         mongoimport -d test -c restaurants --drop --file restaurants.json

Base Setup
~~~~~~~~~~

Create a new Java file called ``AggTour.java`` and include the following import statements:

.. code-block:: java

   import com.mongodb.client.MongoClient;
   import com.mongodb.client.MongoClients;
   import com.mongodb.client.MongoCollection;
   import com.mongodb.client.MongoDatabase;
   import com.mongodb.client.model.Accumulators;
   import com.mongodb.client.model.Aggregates;
   import com.mongodb.client.model.Filters;
   import org.bson.Document;

   import java.util.Arrays;

Connect to a MongoDB Deployment
+++++++++++++++++++++++++++++++

Connect to a MongoDB deployment and declare and define ``MongoDatabase`` and ``MongoCollection`` instances.

> For example, include the following code to connect to a standalone MongoDB deployment running on localhost on port

Reviewer: ```suggestion
Include the following code to connect to a standalone MongoDB deployment running on localhost on port
```

``27017`` and define ``database`` to refer to the ``test`` database and ``collection`` to refer to the ``restaurants``
collection.

.. code-block:: java

   public class AggTour {

       public static void main(String[] args) {
           MongoClient mongoClient = MongoClients.create();
           MongoDatabase database = mongoClient.getDatabase("test");
           MongoCollection<Document> collection = database.getCollection("restaurants");

>            // aggregation here

Reviewer: This was confusing to me initially, but think i figured it out by looking at the example. To make this a little more clear, I might say "// placeholder for your aggregation" and then either "// your connection setup code" or omit completely in the example on line 193.

Author: Okay, I can see how there can be disconnect.

       }
   }

.. tip::

   For information on connecting to MongoDB, see the :ref:`Connection Guide <mongoclient>`

Basic Aggregation Example
~~~~~~~~~~~~~~~~~~~~~~~~~

To perform an aggregation, pass a list of aggregation stages to the `MongoCollection.aggregate() <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/MongoCollection.html#aggregate(com.mongodb.client.ClientSession,java.util.List,java.lang.Class)>`__
method.

The Java driver provides the `Aggregates <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html>`__
helper class that
contains builders for aggregation stages.

In the following example, the aggregation pipeline:

- Uses a :manual:`$match <reference/operator/aggregation/match/>` stage to filter for documents whose
  ``categories`` array field contains the element ``Bakery``. The example uses
  `Aggregates.match <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#match(org.bson.conversions.Bson)>`__
  to build the ``$match`` stage.

> - Uses  a :manual:`$group <reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``

Reviewer: ```suggestion
- Uses a :manual:`$group <reference/operator/aggregation/group/>` stage to group the matching documents by the ``stars``
```

  field, accumulating a count of documents for each distinct value of ``stars``.

.. note::

   The example uses `Aggregates.group <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#group(org.bson.conversions.Bson)>`__
   to build the ``$group`` stage and `Accumulators.sum <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators#sum(java.lang.String,TExpression)>`__
   to build the accumulator expression. The :manual:`accumulator expressions <reference/operator/aggregation/#accumulators-group>`
   used within the ``$group`` stage are built with the `Accumulators <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Accumulators.html>`__

>    helper class.

Reviewer: I think we should use active voice to clearly identify the who and how. E.g. "You can build the accumulator expressions used within the $group stage using the Accumulators helper class. "

Author: I've removed most of this entirely, since it's getting too deep in the weeds and there are going to be other pages that cover these in more detail.
